DataStructures/BitList.cs

```diff
@@ -0,0 +1,1750 @@
+        public void Insert(int index, BitList items)
+        {
+            if (IsEmpty)
+            {
+                bits = new BitArray(items.bits);
+            }
+            else if (index < bits.Count && index >= 0)
+            {
+                var bools = ToBooleanArray().ToList();
+                bools.InsertRange(index, items.ToBooleanArray());
+                bits = new BitArray(bools.ToArray());
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Insert the specified bit in the specified index.
+        /// </summary>
+        public void Insert(int index, bool item)
+        {
+            if (IsEmpty)
+            {
+                bits = new BitArray(new[] { item });
+            }
+            else if (index < bits.Length && index >= 0)
+            {
+                var bools = ToBooleanArray().ToList();
+                bools.Insert(index, item);
+                bits = new BitArray(bools.ToArray());
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Replace a list of bit of the current instance with another list of bit.
+        /// </summary>
+        public void Replace(BitList oldBitList, BitList newBitList)
+        {
+            if (IsEmpty)
+            {
+                throw new EmptyInstanceException();
+            }
+            else
+            {
+                var thisBitListText = ToText();
+                var oldBitListText = oldBitList.ToText();
+                var newBitListText = newBitList.ToText();
+                var bitsText = thisBitListText.Replace(oldBitListText, newBitListText);
+                bits = Parse(bitsText).bits;
+            }
+        }
+
+        /// <summary>
+        /// Write the current instance into a file.
+        /// </summary>
+        public void Write(string filePath)
+        {
+            try
+            {
+                File.WriteAllBytes(filePath, ToByteArray());
+            }
+            catch
+            {
+                throw new FileWritingException();
+            }
+        }
+
+        /// <summary>
+        /// Convert this instance to a text rapresentation.
+        /// </summary>
+        public string ToText()
+        {
+            if (IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            var text = string.Empty;
+            foreach (var bit in ToBooleanArray())
+            {
+                text += bit ? "1" : "0";
+            }
+
+            return text;
+        }
+
+        /// <summary>
+        /// Convert this instance to a specified unmanaged non primitive value array.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public T[] ToUnmanagedNonPrimitiveValueArray<T>() where T : unmanaged
```

Comment:
Codacy found an issue: [The Cyclomatic Complexity of this method is 16 which is greater than 10 authorized.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5559704)